File: app/src/main/java/com/alvinhkh/buseta/ui/MainActivity.java
package com.alvinhkh.buseta.ui;

import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.view.MenuItem;

import com.alvinhkh.buseta.R;
import com.alvinhkh.buseta.follow.dao.FollowDatabase;
import com.alvinhkh.buseta.follow.ui.FollowFragment;
import com.alvinhkh.buseta.mtr.ui.MtrLineStatusFragment;
import com.alvinhkh.buseta.search.ui.HistoryFragment;
import com.alvinhkh.buseta.search.ui.SearchActivity;
import com.alvinhkh.buseta.follow.ui.EditFollowFragment;
import com.alvinhkh.buseta.service.ProviderUpdateService;
import com.alvinhkh.buseta.utils.AdViewUtil;
import com.alvinhkh.buseta.utils.ColorUtil;


public class MainActivity extends BaseActivity {

    private FloatingActionButton fab;

    private BottomNavigationView bottomNavigationView;

    private Fragment followFragment;

    private Fragment historyFragment;

    private Fragment mtrLineStatusFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_follow);

        // set action bar
        setToolbar();
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(R.string.app_name);
            actionBar.setSubtitle(null);
            actionBar.setDisplayHomeAsUpEnabled(false);
        }

        followFragment = new FollowFragment();
        historyFragment = new HistoryFragment();
        mtrLineStatusFragment = MtrLineStatusFragment.newInstance();

        adViewContainer = findViewById(R.id.adView_container);
        if (adViewContainer != null) {
            adView = AdViewUtil.INSTANCE.banner(adViewContainer, adView, false);
        }

        fab = findViewById(R.id.fab);
        fab.setOnClickListener(v -> {
            // if (searchMenuItem != null) searchMenuItem.expandActionView();
            startActivity(new Intent(this, SearchActivity.class));
        });

        bottomNavigationView = findViewById(R.id.bottom_navigation);
        bottomNavigationView.setOnNavigationItemSelectedListener(
                item -> {
                    Integer colorInt = 0;
                    FragmentManager fm = getSupportFragmentManager();
                    if (fm == null) return false;
                    fm.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
                    switch (item.getItemId()) {
                        case R.id.action_follow:
                        {
                            if (fm.findFragmentByTag("follow_list") == null) {
                                if (getSupportActionBar() != null) {
                                    getSupportActionBar().setTitle(R.string.app_name);
                                    getSupportActionBar().setSubtitle(null);
                                }
                                colorInt = ContextCompat.getColor(this, R.color.colorPrimary);
                                FragmentTransaction ft = fm.beginTransaction();
                                ft.replace(R.id.fragment_container, followFragment);
                                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
                                ft.addToBackStack("follow_list");
                                ft.commit();
                            }
                            break;
                        }
                        case R.id.action_search_history:
                        {
                            if (fm.findFragmentByTag("search_history") == null) {
                                if (getSupportActionBar() != null) {
                                    getSupportActionBar().setTitle(R.string.app_name);
                                    getSupportActionBar().setSubtitle(null);
                                }
                                colorInt = ContextCompat.getColor(this, R.color.colorPrimary);
                                FragmentTransaction ft = fm.beginTransaction();
                                ft.replace(R.id.fragment_container, historyFragment);
                                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
                                ft.addToBackStack("search_history");
                                ft.commit();
                            }
                            break;
                        }
                        case R.id.action_railway:
                        {
                            if (fm.findFragmentByTag("railway") == null) {
                                if (getSupportActionBar() != null) {
                                    getSupportActionBar().setTitle(R.string.provider_mtr);
                                    getSupportActionBar().setSubtitle(null);
                                }
                                colorInt = ContextCompat.getColor(this, R.color.provider_mtr);
                                FragmentTransaction ft = fm.beginTransaction();
                                ft.replace(R.id.fragment_container, mtrLineStatusFragment);
                                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
                                ft.addToBackStack("railway");
                                ft.commit();
                            }
                            break;
                        }
                        default:
                            finish();
                    }
                    if (colorInt != 0 && getSupportActionBar() != null) {
                        getSupportActionBar().setBackgroundDrawable(new ColorDrawable(colorInt));
                        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
                                && getWindow() != null) {
                            getWindow().setStatusBarColor(ColorUtil.Companion.darkenColor(colorInt));
                            getWindow().setNavigationBarColor(ColorUtil.Companion.darkenColor(colorInt));
                        }
                    }
                    return true;
                });
        getSupportFragmentManager().addOnBackStackChangedListener(() -> {
            Fragment f = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
            if (f != null && bottomNavigationView != null){
                switch (f.getClass().getName()) {
                    case "FollowFragment":
                        bottomNavigationView.setSelectedItemId(R.id.action_follow);
                        break;
                    case "HistoryFragment":
                        bottomNavigationView.setSelectedItemId(R.id.action_search_history);
                        break;
                    case "MtrLineStatusFragment":
                        bottomNavigationView.setSelectedItemId(R.id.action_railway);
                        break;
                }
            }

        });
        if (bottomNavigationView != null) {
            FollowDatabase followDatabase = FollowDatabase.Companion.getInstance(getApplicationContext());
            if (followDatabase != null && followDatabase.followDao().count() > 0) {
                bottomNavigationView.setSelectedItemId(R.id.action_follow);
            } else {
                bottomNavigationView.setSelectedItemId(R.id.action_search_history);
            }
        }

        try {
            Intent intent = new Intent(this, ProviderUpdateService.class);
            startService(intent);
        } catch (IllegalStateException ignored) {}
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_edit_follow:
                bottomNavigationView.setSelectedItemId(0);
                FragmentManager fragmentManager = getSupportFragmentManager();
                FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
                fragmentTransaction.replace(R.id.fragment_container, new EditFollowFragment());
                fragmentTransaction.addToBackStack("follow_list");
                fragmentTransaction.commit();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (fab != null) {
            fab.show();
        }
    }

    @Override
    protected void onPause() {
        if (fab != null) {
            fab.hide();
        }
        super.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        if (getSupportFragmentManager().getBackStackEntryCount() < 2) {
            finish();
        } else if (getSupportFragmentManager().getBackStackEntryCount() > 0) {
            getSupportFragmentManager().popBackStack();
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/alvinhkh/buseta/ui/MainActivity.kt
package com.alvinhkh.buseta.ui

import android.app.ActivityManager
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import androidx.annotation.ColorInt
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.core.content.ContextCompat

import com.alvinhkh.buseta.R
import com.alvinhkh.buseta.datagovhk.ui.TdTrafficNewsFragment
import com.alvinhkh.buseta.follow.dao.FollowDatabase
import com.alvinhkh.buseta.follow.ui.FollowGroupFragment
import com.alvinhkh.buseta.mtr.ui.MtrLineStatusFragment
import com.alvinhkh.buseta.search.ui.HistoryFragment
import com.alvinhkh.buseta.service.ProviderUpdateService
import com.alvinhkh.buseta.utils.AdViewUtil
import com.alvinhkh.buseta.utils.ColorUtil
import com.google.android.material.snackbar.Snackbar
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.InstallStateUpdatedListener
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.InstallStatus
import com.google.android.play.core.install.model.UpdateAvailability
import timber.log.Timber


class MainActivity : BaseActivity() {

    private val APP_UPDATE_REQUEST_CODE = 1100
    private lateinit var appUpdateManager: AppUpdateManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_follow)
        setToolbar()
        supportActionBar?.run {
            setTitle(R.string.app_name)
            subtitle = null
            setDisplayHomeAsUpEnabled(false)
        }
        adViewContainer = findViewById(R.id.adView_container)
        adView = AdViewUtil.banner(adViewContainer)
        val bottomNavigationView = findViewById<BottomNavigationView>(R.id.bottom_navigation)
        bottomNavigationView.setOnNavigationItemSelectedListener { item ->
            @ColorInt var colorRes = 0
            var title = getString(R.string.app_name)
            val fm = supportFragmentManager
            fm.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
            when (item.itemId) {
                R.id.action_follow -> {
                    if (fm.findFragmentByTag("follow_list") == null) {
                        title = getString(R.string.app_name)
                        colorRes = ContextCompat.getColor(this, R.color.colorPrimary)
                        val ft = fm.beginTransaction()
                        ft.replace(R.id.fragment_container, FollowGroupFragment())
                        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        ft.addToBackStack("follow_list")
                        ft.commit()
                    }
                }
                R.id.action_search_history -> {
                    if (fm.findFragmentByTag("search_history") == null) {
                        title = getString(R.string.app_name)
                        colorRes = ContextCompat.getColor(this, R.color.colorPrimary)
                        val ft = fm.beginTransaction()
                        ft.replace(R.id.fragment_container, HistoryFragment())
                        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        ft.addToBackStack("search_history")
                        ft.commit()
                    }
                }
                R.id.action_railway -> {
                    if (fm.findFragmentByTag("railway") == null) {
                        title = getString(R.string.provider_mtr)
                        colorRes = ContextCompat.getColor(this, R.color.provider_mtr)
                        val ft = fm.beginTransaction()
                        ft.replace(R.id.fragment_container, MtrLineStatusFragment())
                        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        ft.addToBackStack("railway")
                        ft.commit()
                    }
                }
                R.id.action_traffic_news -> {
                    if (fm.findFragmentByTag("traffic_news") == null) {
                        title = getString(R.string.traffic_news)
                        colorRes = Color.parseColor("#008000")
                        val ft = fm.beginTransaction()
                        ft.replace(R.id.fragment_container, TdTrafficNewsFragment())
                        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        ft.addToBackStack("traffic_news")
                        ft.commit()
                    }
                }
                else -> finish()
            }
            supportActionBar?.title = title
            supportActionBar?.subtitle = null
            if (colorRes != 0) {
                supportActionBar?.setBackgroundDrawable(ColorDrawable(colorRes))
                if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
                    window?.statusBarColor = colorRes
                    window?.navigationBarColor = ContextCompat.getColor(this, R.color.bottomNavigationBackground)
                } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    val darkenColor = ColorUtil.darkenColor(colorRes)
                    window?.statusBarColor = darkenColor
                    window?.navigationBarColor = darkenColor
                }
                adViewContainer.setBackgroundColor(colorRes)
                adView = AdViewUtil.banner(adViewContainer, adView, false)
            }
            if (Build.VERSION.SDK_INT >= 28) {
                setTaskDescription(ActivityManager.TaskDescription(title, R.mipmap.ic_launcher,
                        ContextCompat.getColor(this, R.color.colorPrimary600)))
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                val bm = BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher)
                @Suppress("DEPRECATION")
                setTaskDescription(ActivityManager.TaskDescription(title, bm,
                        ContextCompat.getColor(this, R.color.colorPrimary600)))
            }
            true
        }
        supportFragmentManager.addOnBackStackChangedListener {
            val f = supportFragmentManager.findFragmentById(R.id.fragment_container)
            if (f != null && bottomNavigationView != null) {
                when (f.javaClass.name) {
                    "FollowFragment" -> bottomNavigationView.selectedItemId = R.id.action_follow
                    "HistoryFragment" -> bottomNavigationView.selectedItemId = R.id.action_search_history
                    "MtrLineStatusFragment" -> bottomNavigationView.selectedItemId = R.id.action_railway
                    "TdTrafficNewsFragment" -> bottomNavigationView.selectedItemId = R.id.action_traffic_news
                }
            }
        }
        if (bottomNavigationView != null) {
            val followDatabase = FollowDatabase.getInstance(applicationContext)
            if (followDatabase != null && followDatabase.followDao().count() > 0) {
                bottomNavigationView.selectedItemId = R.id.action_follow
            } else {
                bottomNavigationView.selectedItemId = R.id.action_search_history
            }
        }
        try {
            startService(Intent(this, ProviderUpdateService::class.java))
        } catch (ignored: Throwable) {
        }

        val listener = InstallStateUpdatedListener { state ->
            Timber.d("InAppUpdate: state=%s", state)
            if (state.installStatus() == InstallStatus.DOWNLOADED) {
                popupSnackbarForCompleteUpdate()
            }
        }
        appUpdateManager = AppUpdateManagerFactory.create(this)
        appUpdateManager.registerListener(listener)
        appUpdateManager.unregisterListener(listener)
        val appUpdateInfoTask = appUpdateManager.appUpdateInfo
        appUpdateInfoTask.addOnFailureListener { e -> Timber.i(e, "InAppUpdate: error.") }
        appUpdateInfoTask.addOnSuccessListener { appUpdateInfo ->
            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE
                    && appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE)
            ) {
                Timber.i("InAppUpdate: An update is available.")
                appUpdateManager.startUpdateFlowForResult(
                        appUpdateInfo,
                        AppUpdateType.FLEXIBLE,
                        this,
                        APP_UPDATE_REQUEST_CODE)
            } else if (appUpdateInfo.installStatus() == InstallStatus.DOWNLOADED) {
                popupSnackbarForCompleteUpdate()
            }
        }
    }

    override fun onBackPressed() {
        when {
            supportFragmentManager.backStackEntryCount < 2 -> finish()
            supportFragmentManager.backStackEntryCount > 0 -> supportFragmentManager.popBackStack()
            else -> super.onBackPressed()
        }
    }

    override fun onResume() {
        super.onResume()
        appUpdateManager
                .appUpdateInfo
                .addOnSuccessListener { appUpdateInfo ->
                    // If the update is downloaded but not installed,
                    // notify the user to complete the update.
                    if (appUpdateInfo.installStatus() == InstallStatus.DOWNLOADED) {
                        popupSnackbarForCompleteUpdate()
                    }
                }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == APP_UPDATE_REQUEST_CODE) {
            if (resultCode != RESULT_OK) {
                Timber.i("InAppUpdate: Update flow failed! Result code: $resultCode")
            }
        }
    }

    private fun popupSnackbarForCompleteUpdate() {
        Timber.i("InAppUpdate: An update has just been downloaded.")
        Snackbar.make(
                findViewById(android.R.id.content),
                "An update has just been downloaded.",
                Snackbar.LENGTH_INDEFINITE
        ).apply {
            setAction("RESTART") { appUpdateManager.completeUpdate() }
            show()
        }
    }
}
